File: python/seqtree/pairwise.py
"""Needleman-Wunsch and Smith-Waterman: ordinary protein alignment, without BioPython.

Everything else in seqtree *minimises a non-negative penalty* -- that is what a search ball and
an E-value need. This module does the opposite: it **maximises a raw log-odds similarity**, the
way BLAST and BioPython do, because that is what an ordinary pairwise alignment means and what
downstream code expects to get back.

The two views live on the same :class:`~seqtree.SubstitutionMatrix`::

    mat.penalty(a, b)     >= 0, zero on the diagonal   -- search, E-values, gap blocks
    mat.similarity(a, b)  signed log-odds              -- the aligners here

The penalty is the Gram transform of the similarity, ``pen = s(a,a) + s(b,b) - 2·s(a,b)``, which
is **lossy**: it forces the diagonal to zero and destroys ``s(a,a)``. So the raw grid is kept
rather than reconstructed, and a similarity score cannot be recovered from a penalty.

**Conventions, all verified against BioPython** (``tests/python/test_pairwise.py`` runs 6,720
comparisons across three matrices, ten gap settings and both modes; zero disagreements):

* a gap run of length ``L`` costs ``gap_open + (L-1)·gap_extend`` -- ``gap_open`` is the cost of
  the *first* gap column, not a surcharge on top of it;
* ``gap_open == gap_extend`` gives **linear** gaps. There is no separate mode for it;
* ``mode="global"`` charges end gaps like any other (true Needleman-Wunsch, not semi-global);
* ``mode="local"`` never lets the score fall below zero and takes the best cell anywhere
  (Smith-Waterman).

Gap costs are **positive magnitudes** and are subtracted. BLAST's protein defaults are
``gap_open=11, gap_extend=1``; BioPython's ``PairwiseAligner("blastp")`` preset uses ``12, 1``.

Example:
    >>> import seqtree
    >>> from seqtree.pairwise import align, score
    >>> mat = seqtree.SubstitutionMatrix.blosum62()
    >>> score("CASSLGQAYEQYF", "CASSPGQAYEQF", mat)          # global, BLAST defaults
    45
    >>> score("CASSLGQAYEQYF", "CASSPGQAYEQF", mat, gap_open=12)   # BioPython's 'blastp' preset
    44
    >>> aln = align("WWWAAAWWW", "KKKAAAKKK", mat, mode="local")   # Smith-Waterman
    >>> aln.score, aln.aligned_query, aln.aligned_ref
    (12, 'AAA', 'AAA')
"""
from __future__ import annotations

from collections.abc import Sequence

from ._core import Alignment, ScoreMatrix, SubstitutionMatrix
from ._core import align_dist_matrix as _dist_matrix
from ._core import align_pair as _align_pair
from ._core import align_score as _align_score
from ._core import align_score_matrix as _score_matrix

__all__ = ["score", "align", "score_matrix", "dist_matrix"]


def score(
    query: str,
    ref: str,
    matrix: SubstitutionMatrix,
    mode: str = "global",
    gap_open: int = 11,
    gap_extend: int = 1,
    alphabet: str = "aa",
) -> int:
    """Optimal alignment score of ``query`` against ``ref``.

    Args:
        query: First sequence.
        ref: Second sequence.
        matrix: Scoring matrix; its ``similarity`` view is used, not its penalty.
        mode: ``"global"`` for Needleman-Wunsch, ``"local"`` for Smith-Waterman. ``"nw"`` and
            ``"sw"`` are accepted too.
        gap_open: Cost of the first column of a gap. Positive; it is subtracted.
        gap_extend: Cost of each further column. Equal to ``gap_open`` means linear gaps.
        alphabet: ``"aa"``, ``"nt"`` or ``"iupac"``.

    Returns:
        The score, signed. Higher is more similar -- the opposite sense to the rest of seqtree.

    Raises:
        ValueError: On a negative gap cost, an unknown mode, or a symbol outside the alphabet.

    Example:
        >>> m = SubstitutionMatrix.blosum62()
        >>> score("AAA", "AAA", m)
        12
        >>> score("AAA", "AAAAA", m)          # a length-2 gap: 11 + 1*1 = 12
        0
    """
    return _align_score(query, ref, matrix, mode=mode, gap_open=gap_open,
                        gap_extend=gap_extend, alphabet=alphabet)


def align(
    query: str,
    ref: str,
    matrix: SubstitutionMatrix,
    mode: str = "global",
    gap_open: int = 11,
    gap_extend: int = 1,
    alphabet: str = "aa",
) -> Alignment:
    """As :func:`score`, but also returns the aligned strings and the edit ops.

    Returns:
        An ``Alignment``. Note ``Alignment.score`` here is a **similarity** (signed, higher is
        better), whereas the same field from :meth:`seqtree.Index.align` is a penalty. In local
        mode the aligned strings are the matched sub-sequences only.

    Example:
        >>> m = SubstitutionMatrix.blosum62()
        >>> a = align("CASSLGQAYEQYF", "CASSPGQAYEQF", m)
        >>> a.aligned_query, a.aligned_ref
        ('CASSLGQAYEQYF', 'CASSPGQAYEQ-F')
    """
    return _align_pair(query, ref, matrix, mode=mode, gap_open=gap_open,
                       gap_extend=gap_extend, alphabet=alphabet)


def score_matrix(
    queries: Sequence[str],
    refs: Sequence[str],
    matrix: SubstitutionMatrix,
    mode: str = "global",
    gap_open: int = 11,
    gap_extend: int = 1,
    alphabet: str = "aa",
    threads: int = 0,
) -> ScoreMatrix:
    """Every query against every reference, in C++ with the GIL released.

    Returns:
        A :class:`~seqtree.ScoreMatrix` of shape ``(len(queries), len(refs))`` holding signed
        similarity scores. ``numpy.asarray`` wraps it without copying.
    """
    return _score_matrix(list(queries), list(refs), matrix, mode=mode, gap_open=gap_open,
                         gap_extend=gap_extend, alphabet=alphabet, threads=threads)


def dist_matrix(
    queries: Sequence[str],
    refs: Sequence[str],
    matrix: SubstitutionMatrix,
    mode: str = "global",
    gap_open: int = 11,
    gap_extend: int = 1,
    alphabet: str = "aa",
    threads: int = 0,
) -> ScoreMatrix:
    """Alignment **distances**: ``d(a, b) = s(a,a) + s(b,b) - 2·s(a,b)``.

    The Gram transform, applied at the *sequence* level to the alignment scores rather than
    per residue. Non-negative, symmetric, zero on the diagonal -- so it is a distance, and it is
    what a prototype-distance embedding actually wants. This is the quantity users of BioPython
    hand-roll, and it is computed here without a Python loop: the self-scores are taken once per
    sequence, not once per pair.

    Returns:
        A :class:`~seqtree.ScoreMatrix` of shape ``(len(queries), len(refs))``.

    Example:
        >>> m = SubstitutionMatrix.blosum62()
        >>> d = dist_matrix(["CASSLGQAYEQYF"], ["CASSLGQAYEQYF", "CASSPGQAYEQF"], m)
        >>> d[0, 0], d[0, 1] > 0
        (0, True)
    """
    return _dist_matrix(list(queries), list(refs), matrix, mode=mode, gap_open=gap_open,
                        gap_extend=gap_extend, alphabet=alphabet, threads=threads)

File: src/pairwise.cpp
// Needleman-Wunsch (global) and Smith-Waterman (local) with affine gaps, on the raw
// similarity scale. The conventions match Bio.Align.PairwiseAligner exactly, and
// tests/python/test_pairwise.py pins every one of them against BioPython as an oracle.
//
// Gotoh's three-state recurrence, MAXIMISING (everything else in this library minimises a
// non-negative penalty; this does not):
//
//   M[i][j] = s(q_i, r_j) + max(M, X, Y)[i-1][j-1]     both residues aligned
//   X[i][j] = max(M[i-1][j] - open, X[i-1][j] - extend, Y[i-1][j] - open)   query residue vs gap
//   Y[i][j] = max(M[i][j-1] - open, Y[i][j-1] - extend, X[i][j-1] - open)   ref residue vs gap
//
// so a gap run of length L costs open + (L-1)*extend, and open == extend gives linear gaps.
// Local adds a 0 floor on M and takes the best M cell anywhere (Smith-Waterman); the gap
// states are NOT floored -- a local alignment never ends in a gap, you would just trim it.

#include "seqtree/seqtree.hpp"

#include <algorithm>
#include <atomic>
#include <limits>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

namespace seqtree {
namespace {

// Low enough to be unreachable, with headroom so that kNegInf - gap does not overflow.
constexpr int32_t kNegInf = std::numeric_limits<int32_t>::min() / 4;

std::vector<uint8_t> encode(const Codec& c, std::string_view s) {
    std::vector<uint8_t> out(s.size());
    for (size_t i = 0; i < s.size(); ++i) {
        uint8_t v = c.encode(s[i]);
        if (v == Codec::kInvalid)
            throw std::invalid_argument(std::string("invalid symbol '") + s[i] + "' in sequence");
        out[i] = v;
    }
    return out;
}

void check(const SubstitutionMatrix& mat, const Codec& codec, int32_t open, int32_t extend) {
    if (open < 0 || extend < 0)
        throw std::invalid_argument("seqtree: gap_open and gap_extend must be >= 0 (they are "
                                    "magnitudes, and are subtracted)");
    if (mat.size() != codec.size())
        throw std::invalid_argument("matrix size does not match the alphabet");
}

// Cost of a terminal gap run of length L in global mode.
inline int32_t gap_run(int32_t L, int32_t open, int32_t extend) {
    return L <= 0 ? 0 : open + (L - 1) * extend;
}

// Optimal score, two rows per state. O(n) memory, no traceback.
int32_t score_encoded(const uint8_t* q, size_t m, const uint8_t* r, size_t n,
                      const SubstitutionMatrix& mat, AlignMode mode, int32_t go, int32_t ge) {
    const bool local = (mode == AlignMode::Local);
    if (m == 0 || n == 0) {
        if (local) return 0;
        return -gap_run(int32_t(m ? m : n), go, ge);
    }

    std::vector<int32_t> Mp(n + 1), Xp(n + 1), Yp(n + 1);
    std::vector<int32_t> Mc(n + 1), Xc(n + 1), Yc(n + 1);

    // Row 0. Global: the only way to consume ref[0..j) is a leading gap (state Y). Local: an
    // alignment may start anywhere, which the 0 floor on M expresses; the gap states cannot be
    // entered before the alignment has started, so they stay unreachable.
    Mp[0] = 0;
    Xp[0] = kNegInf;
    Yp[0] = kNegInf;
    for (size_t j = 1; j <= n; ++j) {
        Mp[j] = local ? 0 : kNegInf;
        Xp[j] = kNegInf;
        Yp[j] = local ? kNegInf : -gap_run(int32_t(j), go, ge);
    }

    int32_t best = local ? 0 : kNegInf;

    for (size_t i = 1; i <= m; ++i) {
        Mc[0] = local ? 0 : kNegInf;
        Xc[0] = local ? kNegInf : -gap_run(int32_t(i), go, ge);
        Yc[0] = kNegInf;

        for (size_t j = 1; j <= n; ++j) {
            const int32_t s = mat.similarity(q[i - 1], r[j - 1]);
            const int32_t diag = std::max(Mp[j - 1], std::max(Xp[j - 1], Yp[j - 1]));
            int32_t mv = diag + s;
            if (local && mv < 0) mv = 0;
            Mc[j] = mv;

            Xc[j] = std::max(Mp[j] - go, std::max(Xp[j] - ge, Yp[j] - go));
            Yc[j] = std::max(Mc[j - 1] - go, std::max(Yc[j - 1] - ge, Xc[j - 1] - go));

            if (local && mv > best) best = mv;
        }
        Mp.swap(Mc);
        Xp.swap(Xc);
        Yp.swap(Yc);
    }

    if (local) return best;
    return std::max(Mp[n], std::max(Xp[n], Yp[n]));
}

enum : uint8_t { kM = 0, kX = 1, kY = 2, kStop = 3 };

}  // namespace

int32_t align_score(std::string_view query, std::string_view ref, const SubstitutionMatrix& mat,
                    Alphabet alphabet, AlignMode mode, int32_t gap_open, int32_t gap_extend) {
    const Codec codec(alphabet);
    check(mat, codec, gap_open, gap_extend);
    const auto q = encode(codec, query);
    const auto r = encode(codec, ref);
    return score_encoded(q.data(), q.size(), r.data(), r.size(), mat, mode, gap_open, gap_extend);
}

Alignment align_pair(std::string_view query, std::string_view ref, const SubstitutionMatrix& mat,
                     Alphabet alphabet, AlignMode mode, int32_t gap_open, int32_t gap_extend) {
    const Codec codec(alphabet);
    check(mat, codec, gap_open, gap_extend);
    const auto q = encode(codec, query);
    const auto r = encode(codec, ref);
    const int m = int(q.size()), n = int(r.size());
    const bool local = (mode == AlignMode::Local);
    const int32_t go = gap_open, ge = gap_extend;

    Alignment al;
    if (m == 0 || n == 0) {
        al.score = local ? 0 : -gap_run(m ? m : n, go, ge);
        if (!local) {
            for (int i = 0; i < m; ++i) {
                al.aligned_query += query[size_t(i)];
                al.aligned_ref += '-';
                al.ops += 'D';
            }
            for (int j = 0; j < n; ++j) {
                al.aligned_query += '-';
                al.aligned_ref += ref[size_t(j)];
                al.ops += 'I';
            }
        }
        return al;
    }

    const int W = n + 1;
    const size_t sz = size_t(m + 1) * W;
    std::vector<int32_t> M(sz, kNegInf), X(sz, kNegInf), Y(sz, kNegInf);
    std::vector<uint8_t> bM(sz, kStop), bX(sz, kM), bY(sz, kM);

    M[0] = 0;
    for (int j = 1; j <= n; ++j) {
        M[j] = local ? 0 : kNegInf;
        Y[j] = local ? kNegInf : -gap_run(j, go, ge);
        bY[j] = (j > 1) ? kY : kM;
    }
    for (int i = 1; i <= m; ++i) {
        const size_t k = size_t(i) * W;
        M[k] = local ? 0 : kNegInf;
        X[k] = local ? kNegInf : -gap_run(i, go, ge);
        bX[k] = (i > 1) ? kX : kM;
    }

    // argmax over the three states, reporting which one won.
    auto pick = [](int32_t vm, int32_t vx, int32_t vy, uint8_t& who) {
        int32_t v = vm;
        who = kM;
        if (vx > v) { v = vx; who = kX; }
        if (vy > v) { v = vy; who = kY; }
        return v;
    };

    int32_t best = local ? 0 : kNegInf;
    int bi = 0, bj = 0;

    for (int i = 1; i <= m; ++i) {
        for (int j = 1; j <= n; ++j) {
            const size_t k = size_t(i) * W + j;
            const size_t kd = size_t(i - 1) * W + (j - 1);
            const size_t ku = size_t(i - 1) * W + j;
            const size_t kl = size_t(i) * W + (j - 1);

            uint8_t who;
            int32_t mv = pick(M[kd], X[kd], Y[kd], who) + mat.similarity(q[i - 1], r[j - 1]);
            if (local && mv < 0) { mv = 0; who = kStop; }
            M[k] = mv;
            bM[k] = who;

            X[k] = pick(M[ku] - go, X[ku] - ge, Y[ku] - go, bX[k]);
            Y[k] = pick(M[kl] - go, X[kl] - go, Y[kl] - ge, bY[k]);

            if (local && mv > best) { best = mv; bi = i; bj = j; }
        }
    }

    int i, j;
    uint8_t st;
    if (local) {
        al.score = best;
        i = bi;
        j = bj;
        st = kM;
        if (best == 0) return al;  // no positive-scoring local alignment exists
    } else {
        const size_t end = size_t(m) * W + n;
        al.score = pick(M[end], X[end], Y[end], st);
        i = m;
        j = n;
    }

    while (i > 0 || j > 0) {
        const size_t k = size_t(i) * W + j;
        if (st == kM) {
            if (i == 0 || j == 0) break;
            if (local && M[k] == 0) break;  // Smith-Waterman stops where the score resets
            const char a = codec.decode(q[i - 1]), b = codec.decode(r[j - 1]);
            al.aligned_query += a;
            al.aligned_ref += b;
            al.ops += (a == b ? 'M' : 'S');
            st = bM[k];
            --i;
            --j;
            if (st == kStop) break;
        } else if (st == kX) {
            if (i == 0) break;
            al.aligned_query += codec.decode(q[i - 1]);
            al.aligned_ref += '-';
            al.ops += 'D';
            st = bX[k];
            --i;
        } else {
            if (j == 0) break;
            al.aligned_query += '-';
            al.aligned_ref += codec.decode(r[j - 1]);
            al.ops += 'I';
            st = bY[k];
            --j;
        }
    }
    std::reverse(al.aligned_query.begin(), al.aligned_query.end());
    std::reverse(al.aligned_ref.begin(), al.aligned_ref.end());
    std::reverse(al.ops.begin(), al.ops.end());
    return al;
}

namespace {

// Shared driver for the two N*K matrices. `dist` applies the sequence-level Gram transform
// d = s(a,a) + s(b,b) - 2*s(a,b); the self-scores are computed once per sequence, not per pair.
std::vector<int32_t> matrix_impl(const std::vector<std::string>& queries,
                                 const std::vector<std::string>& refs,
                                 const SubstitutionMatrix& mat, Alphabet alphabet, AlignMode mode,
                                 int32_t go, int32_t ge, int threads, bool dist) {
    const Codec codec(alphabet);
    check(mat, codec, go, ge);

    const size_t N = queries.size(), K = refs.size();
    std::vector<int32_t> out(N * K);
    if (N == 0 || K == 0) return out;

    std::vector<std::vector<uint8_t>> qc(N), rc(K);
    for (size_t i = 0; i < N; ++i) qc[i] = encode(codec, queries[i]);
    for (size_t k = 0; k < K; ++k) rc[k] = encode(codec, refs[k]);

    std::vector<int32_t> q_self(N), r_self(K);
    if (dist) {
        for (size_t i = 0; i < N; ++i)
            q_self[i] = score_encoded(qc[i].data(), qc[i].size(), qc[i].data(), qc[i].size(), mat,
                                      mode, go, ge);
        for (size_t k = 0; k < K; ++k)
            r_self[k] = score_encoded(rc[k].data(), rc[k].size(), rc[k].data(), rc[k].size(), mat,
                                      mode, go, ge);
    }

    unsigned nt =
        threads > 0 ? unsigned(threads) : std::max(1u, std::thread::hardware_concurrency());
    nt = std::min<unsigned>(nt, std::max<size_t>(1, N));

    std::atomic<size_t> next{0};
    auto worker = [&] {
        for (;;) {
            const size_t i = next.fetch_add(1);
            if (i >= N) break;
            int32_t* row = out.data() + i * K;
            for (size_t k = 0; k < K; ++k) {
                const int32_t s = score_encoded(qc[i].data(), qc[i].size(), rc[k].data(),
                                                rc[k].size(), mat, mode, go, ge);
                row[k] = dist ? (q_self[i] + r_self[k] - 2 * s) : s;
            }
        }
    };

    std::vector<std::thread> pool;
    for (unsigned t = 0; t < nt; ++t) pool.emplace_back(worker);
    for (auto& th : pool) th.join();
    return out;
}

}  // namespace

std::vector<int32_t> align_score_matrix(const std::vector<std::string>& queries,
                                        const std::vector<std::string>& refs,
                                        const SubstitutionMatrix& mat, Alphabet alphabet,
                                        AlignMode mode, int32_t gap_open, int32_t gap_extend,
                                        int threads) {
    return matrix_impl(queries, refs, mat, alphabet, mode, gap_open, gap_extend, threads, false);
}

std::vector<int32_t> align_dist_matrix(const std::vector<std::string>& queries,
                                       const std::vector<std::string>& refs,
                                       const SubstitutionMatrix& mat, Alphabet alphabet,
                                       AlignMode mode, int32_t gap_open, int32_t gap_extend,
                                       int threads) {
    return matrix_impl(queries, refs, mat, alphabet, mode, gap_open, gap_extend, threads, true);
}

}  // namespace seqtree
